### PR TITLE
Add private ip address validity check

### PR DESCRIPTION
**Pull request by minsulee2**
Updates:
- Add isValidPrivateUrl function on rule-util and common-util
- Add test case

Related issues:
- https://github.com/ainblockchain/ain-blockchain/issues/699

### CODE REVIEW

File: p2p/index.js [hunk 1]
```diff
@@ -561,6 +561,27 @@ class P2pClient {
...
+   */
+  isValidUrl(url) {
+    return HOSTING_ENV === 'local' ? CommonUtil.isValidPrivateUrl(url) : CommonUtil.isValidUrl(url);
+  }
+
+  /**
+   * Returns only private url without 'http://', '/json-rpc', and 'port number'.
+   * @param {*} url is an IPv4 ip address with protocols.
+   */
+  getOnlyPrivateUrl(url) {
```
Comment (platfowner) [inline p2p/index.js:576]: what about using a public package like url-parse?

Comment (platfowner) [inline p2p/index.js:576]: what about 'getHostnameOfUrl'?

File: p2p/index.js [hunk 2]
```diff
@@ -561,6 +561,27 @@ class P2pClient {
...
+   * Checks validity of url based on HOSTING_ENV.
+   * @param {string} url is an IPv4 ip address.
+   */
+  isValidUrl(url) {
+    return HOSTING_ENV === 'local' ? CommonUtil.isValidPrivateUrl(url) : CommonUtil.isValidUrl(url);
+  }
+
+  /**
+   * Returns only private url without 'http://', '/json-rpc', and 'port number'.
+   * @param {*} url is an IPv4 ip address with protocols.
```
Comment (platfowner) [inline p2p/index.js:574]: s/*/string/?

File: p2p/index.js [hunk 3]
```diff
@@ -586,8 +607,8 @@ class P2pClient {
     this.peerCandidates[peerCandidateJsonRpcUrl] = { queriedAt: Date.now() };
     const peerCandidateUrlList = _.get(peerCandidateInfo, 'peerCandidateUrlList', []);
     peerCandidateUrlList.forEach(url => {
-      // FIXME(minsulee2): CommonUtil.isValidUrl is not working for internal ip addresses.
-      if (!this.peerCandidates[url] && CommonUtil.isValidUrl(url)) {
+      const onlyPrivateUrl = this.getOnlyPrivateUrl(url);
```
Comment (platfowner) [inline p2p/index.js:610]: why do we need the hostname part only?

File: db/rule-util.js [hunk 4]
```diff
@@ -319,6 +319,11 @@ class RuleUtil {
     return this.isString(url) ? strictUrlRegex.test(url) : false;
   }
 
+  isValidPrivateUrl(url) {
+    const privateUrlRegex = /(^127\.)|(^192\.168\.)|(^10\.)|(^172\.1[6-9]\.)|(^172\.2[0-9]\.)|(^172\.3[0-1]\.)|(^::1$)|(^[fF][cCdD])/;
```
Comment (platfowner) [inline db/rule-util.js:323]: it seems there are some public packages, e.g. https://www.npmjs.com/package/private-ip
aren't they not applicable?

File: db/rule-util.js [hunk 5]
```diff
@@ -319,6 +319,11 @@ class RuleUtil {
     return this.isString(url) ? strictUrlRegex.test(url) : false;
   }
 
+  isValidPrivateUrl(url) {
```
Comment (platfowner) [inline db/rule-util.js:322]: if it works only for ip addresses, what about naming it isValidPrivateIp?

File: p2p/index.js (hunk 1 above)
Comment (minsulee2) [reply p2p/index.js]: Removed this function!

Comment (minsulee2) [reply p2p/index.js]: Removed!

File: p2p/index.js (hunk 2 above)
Comment (minsulee2) [reply p2p/index.js]: This function is removed!

File: p2p/index.js (hunk 3 above)
Comment (minsulee2) [reply p2p/index.js]: There was no need for this! I updated it.

File: db/rule-util.js (hunk 4 above)
Comment (minsulee2) [reply db/rule-util.js]: Thought we do not use external dependencies for rule-tuil as much as possible, do we?

File: db/rule-util.js (hunk 5 above)
Comment (minsulee2) [reply db/rule-util.js:322]: Updated!

**Comment by minsulee2** [comment]
PTAL. Sorry for poor dev for this. I must've been too tired. Updated regex and isValidJsonRpcUrl function!

File: db/rule-util.js (hunk 4 above)
Comment (platfowner) [reply db/rule-util.js]: You're right. I was thinking that the function can be added to only CommonUtil, but it looks just fine now.

File: unittest/rule-util.test.js [hunk 6]
```diff
@@ -283,6 +283,35 @@ describe("RuleUtil", () => {
...
+      expect(util.isValidPrivateUrl('0x6af1ec8d4f0a55bac328cb20336ed0eff46fa6334ebd112147892f1b15aafc8')).to.equal(false);
+      expect(util.isValidPrivateUrl('ainetwork.ai')).to.equal(false);
+      expect(util.isValidPrivateUrl('https://*.ainetwork.ai')).to.equal(false);
+      expect(util.isValidPrivateUrl('http://172.16.0.36:8080/json-rpc')).to.equal(false);
+    })
+
+    it("when valid input", () => {
+      expect(util.isValidPrivateUrl('172.16.0.36')).to.equal(true);
+      expect(util.isValidPrivateUrl('http://172.16.0.36')).to.equal(true);
+      expect(util.isValidPrivateUrl('http://172.16.0.36:8080')).to.equal(true);
```
Comment (platfowner) [inline unittest/rule-util.test.js:311]: some https test cases with and without port numbers?

File: p2p/index.js [hunk 7]
```diff
@@ -561,6 +561,20 @@ class P2pClient {
     }, WAIT_FOR_ADDRESS_TIMEOUT_MS);
   }
 
+  /**
+   * Checks validity of JSON-RPC endpoint url based on HOSTING_ENV.
+   * @param {string} url is an IPv4 ip address.
+   */
+  isValidJsonRpcUrl(url) {
+    const urlWithoutJsonRpc = url.includes('/json-rpc') ? url.substr(0, url.length - 9) : false;
```
Comment (platfowner) [inline p2p/index.js:569]: could you add type check CommonUtil.isString()?
and for '/json-rpc' what about defining a constant variable, e.g. 
```
  const JSON_RPC_PATH = '/json-rpc';
```
and using endsWith() instead? e.g.
```
  ... = url.endsWith(JSON_RPC_PATH) ? url.slice(0, -JSON_RPC_PATH.length) : false;
```

File: unittest/rule-util.test.js (hunk 6 above)
Comment (minsulee2) [reply unittest/rule-util.test.js:312]: Updated!

**Comment by minsulee2** [comment]
PTAL!

File: p2p/index.js [hunk 8]
```diff
@@ -561,6 +561,25 @@ class P2pClient {
...
+   * Checks validity of JSON-RPC endpoint url based on HOSTING_ENV.
+   * @param {string} url is an IPv4 ip address.
+   */
+  isValidJsonRpcUrl(url) {
+    if (!CommonUtil.isString(url)) {
+      return false;
+    }
+    const JSON_RPC_PATH = '/json-rpc';
+    const urlWithoutJsonRpc =
+        url.endsWith(JSON_RPC_PATH) ? url.slice(0, JSON_RPC_PATH.length) : false;
```
Comment (platfowner) [inline p2p/index.js:574]: isn't '-' missing? (-JSON_RPC_PATH.length)

Comment (minsulee2) [reply p2p/index.js]: True! Updated and Test done! Thanks!!

**Comment by minsulee2** [comment]
Sorry for the tiny mistake. PTAL!

**Review by platfowner — APPROVED** [review]
LGTM
Thanks for the nice code.

**Comment by minsulee2** [comment]
Thanks! Merging now!

**Pull request merged**